x-pack/plugin/blob-cache/src/test/java/org/elasticsearch/blobcache/shared/SharedBlobCacheServiceTests.java

```diff
@@ -104,7 +104,7 @@ public void testBasicEviction() throws IOException {
                 ByteRange.of(0L, 1L),
                 ByteRange.of(0L, 1L),
                 (channel, channelPos, relativePos, length) -> 1,
-                (channel, channelPos, relativePos, length, progressUpdater) -> progressUpdater.accept(length),
+                (channel, channelPos, streamFactory, relativePos, length, progressUpdater) -> progressUpdater.accept(length),
```

Comment:
nit I wonder if a simple unit test would also be useful in this core elasticsearch PR? Maybe a unit test that pre-fills some gaps in a region, then calls the populateAndRead(), and makes sure that the SourceInputStreamFactory.create() is called in the expected sequential order as many times as the holes.

Reply:
Thanks, it's a good point on adding a test. I have pushed 8244f881122bbe88a3484f19959803673548becb

> makes sure that the SourceInputStreamFactory.create() is called in the expected sequential order as many times as the holes.

This is not viable since whether to call the method is up to the implementation. But we can assert a similar thing that is `fillCacheRange` is called as many times as the number of holes and the invocation must be in a single thread and in gap order. The test also ensures the shared factory object is passed to each invocation. These are the behaviours a downstream implementation would rely on. I hope this makes sense.